### PR TITLE
UpdateGrpcCmd#4656

### PR DESCRIPTION
**Pull request by Abhi132004**
To provide better error messages when the grpc command crashes, I have also considered improving the error handling within the ''generateBalFile'' method and other relevant methods in the same Java file. Look for any potential exceptions that may occur during protocol buffer processing and handle them appropriately by providing informative error messages.

Isse: https://github.com/ballerina-platform/ballerina-standard-library/issues/4656

## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests
- [x] Updated the spec

### CODE REVIEW

File: protoc-cli/src/main/java/io/ballerina/protoc/protobuf/cmd/GrpcCmd.java [hunk 1]
```diff
@@ -125,46 +125,38 @@ private static void exportResource(String resourceName, ClassLoader classLoader)
...
             if (pathExtension.isEmpty() || !PROTO_EXTENSION.equalsIgnoreCase(pathExtension.get())) {
-                String errorMessage = "Invalid proto file path. Please input valid proto file location.";
-                outStream.println(errorMessage);
-                return;
+                throw new RuntimeException("Invalid proto file path. Please input a valid proto file location."); // Improved error message
             }
             generateBalFile(protoPath);
         }
+    } catch (Exception e) {
+        // Print a user-friendly error message
```
Comment (gayaldassanayake) [inline protoc-cli/src/main/java/io/ballerina/protoc/protobuf/cmd/GrpcCmd.java]: I feel like this comment is not required. The error message printing logic in L156 is self explainatory.

File: protoc-cli/src/main/java/io/ballerina/protoc/protobuf/cmd/GrpcCmd.java [hunk 2]
```diff
@@ -125,46 +125,38 @@ private static void exportResource(String resourceName, ClassLoader classLoader)
...
+        outStream.println("Error: " + e.getMessage());
     }
+}
+
```
Comment (gayaldassanayake) [inline protoc-cli/src/main/java/io/ballerina/protoc/protobuf/cmd/GrpcCmd.java:160]:
```suggestion
```

File: protoc-cli/src/main/java/io/ballerina/protoc/protobuf/cmd/GrpcCmd.java [hunk 3]
```diff
@@ -125,46 +125,38 @@ private static void exportResource(String resourceName, ClassLoader classLoader)
...
             }
         } else {
             // Single proto file
-            // check input protobuf file path
             Optional<String> pathExtension = getFileExtension(protoPath);
             if (pathExtension.isEmpty() || !PROTO_EXTENSION.equalsIgnoreCase(pathExtension.get())) {
-                String errorMessage = "Invalid proto file path. Please input valid proto file location.";
-                outStream.println(errorMessage);
-                return;
+                throw new RuntimeException("Invalid proto file path. Please input a valid proto file location."); // Improved error message
```
Comment (gayaldassanayake) [inline protoc-cli/src/main/java/io/ballerina/protoc/protobuf/cmd/GrpcCmd.java]:
```suggestion
                throw new RuntimeException("Invalid proto file path. Please input a valid proto file location.");
```

File: protoc-cli/src/main/java/io/ballerina/protoc/protobuf/cmd/GrpcCmd.java [hunk 4]
```diff
@@ -125,46 +125,38 @@ private static void exportResource(String resourceName, ClassLoader classLoader)
...
-            try {
-                protoFiles = getProtoFiles(protoPath);
-            } catch (IOException e) {
-                String errorMessage = "Failed to find proto files in the directory. " +
-                        "Please input a valid proto files directory.";
-                outStream.println(errorMessage);
-                return;
+            List<String> protoFiles = getProtoFiles(protoPath);
+            if (protoFiles.isEmpty()) {
+                throw new RuntimeException("Input directory does not contain any proto files. Please input a valid proto files directory."); // Improved error message
```
Comment (gayaldassanayake) [inline protoc-cli/src/main/java/io/ballerina/protoc/protobuf/cmd/GrpcCmd.java]:
```suggestion
                throw new RuntimeException("Input directory does not contain any proto files. Please input a valid proto files directory.");
```

Also if the line length is >120 characters, you may consider breaking it into multiple lines (Applicable for the other places as well).

File: protoc-cli/src/main/java/io/ballerina/protoc/protobuf/cmd/GrpcCmd.java (hunk 1 above)
Comment (Abhi132004) [reply protoc-cli/src/main/java/io/ballerina/protoc/protobuf/cmd/GrpcCmd.java]: sir, @gayaldassanayake  changes implemented and new PR as a patch2 opened successfully.

**Review by Abhi132004 — COMMENTED** [review]
@gayaldassanayake  Suggested changes have been implemented and made a new PR in the same issue , requested to review

File: protoc-cli/src/main/java/io/ballerina/protoc/protobuf/cmd/GrpcCmd.java [hunk 5]
```diff
@@ -125,46 +125,36 @@ private static void exportResource(String resourceName, ClassLoader classLoader)
...
-            try {
-                protoFiles = getProtoFiles(protoPath);
-            } catch (IOException e) {
-                String errorMessage = "Failed to find proto files in the directory. " +
-                        "Please input a valid proto files directory.";
-                outStream.println(errorMessage);
-                return;
+            List<String> protoFiles = getProtoFiles(protoPath);
+            if (protoFiles.isEmpty()) {
+                throw new RuntimeException("Input directory does not contain any proto files. Please input a valid proto files directory.");
```
Comment (ThisaruGuruge) [inline protoc-cli/src/main/java/io/ballerina/protoc/protobuf/cmd/GrpcCmd.java]: This line exceeds the character limit. (The limit is 120 characters per line).

Comment (Abhi132004) [reply protoc-cli/src/main/java/io/ballerina/protoc/protobuf/cmd/GrpcCmd.java]: Issue resolved, patch 3 pr opened in the same for the same, please check: https://github.com/ballerina-platform/ballerina-standard-library/issues/4656

**Pull request closed without merging**